<a href="https://colab.research.google.com/github/akashf/DataScience/blob/master/DeepLearingLab_Linear_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code.

<h4 align="center">Downloading the data set</h4>

In [0]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

In [3]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [5]:
concrete_data.shape

(1030, 9)

In [7]:
concrete_data.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [8]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

#### Split data into predictors and target
The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [0]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
target = concrete_data['Strength']

In [11]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [12]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [13]:
predictors_norm = (predictors - predictors.mean())/predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [15]:
n_cols = predictors_norm.shape[1]
n_cols

8

#Import Keras

Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.

In [16]:
import keras

Using TensorFlow backend.


In [0]:
from keras.models import Sequential
from keras.layers import Dense

## Build a Neural Network
Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [0]:
#def regression model
def regression_model():
  #create model
  model = Sequential()
  model.add(Dense(50, activation = 'relu', input_shape=(n_cols,)))
  model.add(Dense(50, activation = 'relu'))
  model.add(Dense(1))
  
  #compile model
  model.compile(optimizer='adam', loss = 'mean_squared_error')
  return model

The above function create a model that has two hidden layers, each of 50 hidden units.

## Train and Test the Network

Let's call the function now to create our model.

In [0]:
#build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.

In [21]:
#fit the model
model.fit(predictors_norm,target, validation_split=0.3,epochs=100,verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 0s - loss: 1619.5568 - val_loss: 1105.3520
Epoch 2/100
 - 0s - loss: 1480.0734 - val_loss: 964.7648
Epoch 3/100
 - 0s - loss: 1246.1773 - val_loss: 757.8646
Epoch 4/100
 - 0s - loss: 908.9902 - val_loss: 507.4155
Epoch 5/100
 - 0s - loss: 540.6459 - val_loss: 293.4986
Epoch 6/100
 - 0s - loss: 302.0012 - val_loss: 200.4964
Epoch 7/100
 - 0s - loss: 232.0977 - val_loss: 181.0801
Epoch 8/100
 - 0s - loss: 216.2727 - val_loss: 175.5867
Epoch 9/100
 - 0s - loss: 205.3116 - val_loss: 172.9541
Epoch 10/100
 - 0s - loss: 197.2587 - val_loss: 174.9807
Epoch 11/100
 - 0s - loss: 190.2597 - val_loss: 170.5384
Epoch 12/100
 - 0s - loss: 184.2149 - val_loss: 167.3372
Epoch 13/100
 - 0s - loss: 179.5322 - val_loss: 170.7634
Epoch 14/100
 - 0s - loss: 174.8418 - val_loss: 169.7562
Epoch 15/100
 - 0s - loss: 170.9602 - val_loss: 165.4903
Epoch 16/100
 - 0s - loss: 167.4258 - val_loss: 169.7839
Epoch 17/100
 - 0s - loss: 164.2483 - val_loss:

Feel free to vary the following and note what impact each change has on the model's performance:

1. Increase or decreate number of neurons in hidden layers
2. Add more hidden layers
3. Increase number of epochs